### ISSUE TITLE
Use dev services by default in Quarkus' own integration tests

### ISSUE THREAD

**Issue by yrodiere** [quarkusio/quarkus#44124]
### Description

Test modules in `integration-tests` rely on [dedicated Maven configuration](https://github.com/quarkusio/quarkus/blob/c679d63747a2656396d2aaf54e43d6277a5e5324/integration-tests/elasticsearch-rest-client/pom.xml#L168-L235) to start containers of services used in tests.

Could we not just rely on dev services for most of them?

One benefit would be additional testing of dev services in scenarios that are as close as possible to the "real world" -- as opposed to most dev service tests we have right now, which rely on `QuarkusDevModeTest` -- which, as great as it is, is still a simulated setup.

Another benefit in our own dev environments would be that a test failure in surefire wouldn't leave the containers running, forcing us to manually kill them before restarting the tests (I swear I had to do this many times).

A downside could be performance, as we would start the containers twice per integration test module (once for surefire tests, once for failsafe tests) instead of once. That can be worked around on CI by enabling container reuse, though then cleanup mid-build will be challenging.

Related: #43980. Which could, perhaps, have been caught if we used dev services in our integration tests. But I think the points above stand even if it's a bad example.

**Comment by yrodiere** [comment]
> [@stuartwdouglas](https://github.com/stuartwdouglas) (devservices)

We might want to update the bot config...

**Comment by geoand** [comment]
> Test modules in integration-tests rely on [dedicated Maven configuration](https://github.com/quarkusio/quarkus/blob/c679d63747a2656396d2aaf54e43d6277a5e5324/integration-tests/elasticsearch-rest-client/pom.xml#L168-L235) to start containers of services used in tests.

For a few things there might be specific reasons why Testcontainers were not used, but I believe that most cases are just leftovers from when we didn't have DevServices

**Comment by gsmet** [comment]
Yeah, same here. For MongoDB, I remember we have to avoid Dev Services because we are using clustering features. But apart from that, I think we could move to Dev Services.

**Comment by yrodiere** [comment]
> Note that there's one important limitation though: Windows CI doesn't support containers. So that means we need to keep not enabling the tests and not starting dev services when containers are not enabled.

Looks like an argument in favor of dev services to me: we'll need to disable the tests (same as today), but now disabling the tests will automatically disable container usage, since dev services only start when you start the tests :)

**Comment by holly-cummins** [comment]
I don't know the background of why we weren't using dev services, so can't comment with authority, but I agree that having less home-rolled code in our test poms, in combination with more coverage of dev services, can only be a good thing. For things like catching the next https://github.com/quarkusio/quarkus/issues/43980 we'd also probably need a regular Mac CI, but that's also something we want to do anyway.

**Comment by gsmet** [comment]
> I don't know the background of why we weren't using dev services

There is a mix of they weren't available at that time or flexible enough and I think in one case (some MongoDB testing), we actually still need clustering support that is easier to handler when starting MongoDB manually.

So I think it's worth doing when actually doable.

**Comment by yrodiere** [comment]
FWIW, the current setup (Maven starts containers) has repeatedly bitten us when onboarding new contributors, where they had a hard time running JPA/Hibernate integration tests from the IDE.

Some cases are even worse because of misconfiguration: two datasources, one with a JDBC URL pointing to the Maven-started container, and the other one used for testing of Hibernate ORM offline startup that had no JDBC URL. When started from the IDE, you get an exception for everything accessing the first datasource (Maven didn't start the container) but you still see a container started in podman (the second datasource has no JDBC URL so devservices kick in). A bit confusing :)

So I'll try to prioritize this...

**Comment by holly-cummins** [comment]
I will start chipping away at this to give me dev services coverage while I convert the datasource dev services to the new model.

**Comment by yrodiere** [comment]
Thank you!

FWIW you might want to stay away from Hibernate Search tests while @marko-bekhta is figuring out the problems with Elasticsearch dev services.
But if you focus on datasources, I don't think Hibernate Search will be a problem -- IIRC it's mostly tested against H2.

**Comment by holly-cummins** [comment]
Another potential risk of doing this is a loss of coverage of non-dev-service scenarios. @yrodiere very sensibly suggests a solution here is 

> Personally it's a risk I'm willing to take, but if we want to be extra careful... It could be enough to have one single test in each DB extension where we disable dev services, use testcontainers explicitly with an explicit port and check that we can connect to the DB?

I'll track progress across the datasources here, since smaller changes will be easier to review than one huge one (and I want to make sure I'm on the right track before I take a machete to the poms). 

- [X] Oracle, https://github.com/quarkusio/quarkus/pull/51059 (new tests not added, since `extensions/reactive-oracle-client` didn't get converted to dev services and therefore gives coverage)
- [X] Postgresql https://github.com/quarkusio/quarkus/pull/51066
- [X] MariaDB https://github.com/quarkusio/quarkus/pull/51090
- [X] MySQL https://github.com/quarkusio/quarkus/pull/51092
- [X] MSSQL https://github.com/quarkusio/quarkus/pull/51093
- [X] DB2 https://github.com/quarkusio/quarkus/pull/51128

Derby and H2 should be a no-op.

**Comment by holly-cummins** [comment]
## Follow-on items
- Get rid of other uses of the docker maven plugin where dev services don't make sense for coverage reasons, or are just hard (#51125)
- Re-introduce some of the careful performance optimising we'd done for the database containers (#51126)
